src/Tools/ExternalAccess/Razor/Cohost/RazorCohostDidChangeEndpoint.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+
+[LanguageServerEndpoint(Methods.TextDocumentDidChangeName)]
+[ExportRazorStatelessLspService(typeof(RazorCohostDidChangeEndpoint)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class RazorCohostDidChangeEndpoint([Import(AllowDefault = true)] IRazorCohostDidChangeHandler? didChangeHandler) : ILspServiceDocumentRequestHandler<DidChangeTextDocumentParams, object?>
+{
+    public bool MutatesSolutionState => true;
+    public bool RequiresLSPSolution => false;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(DidChangeTextDocumentParams request)
+        => request.TextDocument;
+
+    public async Task<object?> HandleRequestAsync(DidChangeTextDocumentParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        var text = context.GetTrackedDocumentSourceText(request.TextDocument.Uri);
+
+        // Per the LSP spec, each text change builds upon the previous, so we don't need to translate any text
+        // positions between changes, which makes this quite easy. See
+        // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#didChangeTextDocumentParams
+        // for more details.
+        foreach (var change in request.ContentChanges)
+            text = text.WithChanges(ProtocolConversions.ContentChangeEventToTextChange(change, text));
+
+        context.UpdateTrackedDocument(request.TextDocument.Uri, text);
+
+        // Razor can't handle this request because they don't have access to the RequestContext, but they might want to do something with it
+        if (didChangeHandler is not null)
```

Comment:
Wondering here if we want a different mechanism to expose changes to Razor, if Razor needs to do special handling on didChange.  For example similar to the IOnInitialized services that get called by the base InitializedHandler.

Especially when we unify with the Roslyn server- I'm not sure we want to allow separate didChange handlers per language. 

Does the work Razor needs to do here require the queue to be blocked while its happening?

Same for open/close

Reply:
All good thoughts, and I see this as definitely a point-in-time thing.

1. When we have dynamic registration, yes I would expect the "normal" text document sync handlers work for all languages and we'd remove these.
2. When Razor can pull on the source generator for our .g.cs files then we won't need to be notified about changes or opens.

For now though, we need something here, as the Razor side that responds to this is effectively unchanged from our existing code, we kick off a compile of the razor file as it changes.

Reply:
Actually, even when we can use the source generator and even use the Roslyn API rather than LSP to ask questions about it etc, we'd still need some way to get notified of open and change of Razor files, and run code while the queue is blocked, I order to make sure the html virtual buffers are up to date. Maybe some way to hook into the LspWorkspaceManager? Haven't really thought about the specifics of that far ahead yet.

Reply:
Hmm.  Probably depends on what needs to happen.  I wouldn't be opposed to something like IOnInitialized where we expose a hook inside the handler.  I think putting it there makes it more clear that its a blocking hook (than putting in LspWorkspaceManager0

Reply:
I'm missing something. Isn't that exactly what this is currently doing?

Reply:
What I meant was putting code inside the main DidChangeHandler that allows for arbitrary additional callers, instead of duplicating teh handler for Razor and having a special Razor hook.
e.g. https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/Handler/ServerLifetime/InitializedHandler.cs#L27

We'd have a generic hook that razor could implement.  However I think you need to export a separate Razor didChange handler already - so I'm fine if we do that later when we merge the servers.

Reply:
Oh yep, can definitely follow up with that once we merge. The existing handlers won't activate in this server though, due to the export attribute being used.